src/cryptoadvance/specter/global_search.py

```diff
@@ -0,0 +1,260 @@
+import os
+import logging
+
+from flask import url_for
+
+logger = logging.getLogger(__name__)
+
+
+class HtmlElement:
+    "This is a way to reconstruct the HTML Logical UI Tree and sum up results nicely"
+
+    def __init__(
+        self,
+        parent,
+        id=None,
+        function=None,
+        children=None,
+        visible_on_endpoints="/",
+        filter_via_input_ids=None,
+    ):
+        self.parent = parent
+        if self.parent:
+            self.parent.children.add(self)
+        self.children = children if children else set()
+        self.id = id if id else set()
+        self._result = None
+        self.function = function
+        self.visible_on_endpoints = visible_on_endpoints
+        self.filter_via_input_ids = filter_via_input_ids
+
+    @property
+    def result(self):
+        if self._result:
+            return self._result
+        else:
+            return sum([child.result for child in self.children])
+
+    @result.setter
+    def result(self, result):
+        if self.children:
+            raise "Setting results is only allowed for end nodes"
+        self._result = result
+
+    def calculate_end_nodes(self):
+        if not self.children:
+            return [self]
+
+        end_nodes = []
+        for child in self.children:
+            end_nodes += child.calculate_end_nodes()
+        return end_nodes
+
+    def reset_results(self):
+        end_nodes = self.calculate_end_nodes()
+        for node in end_nodes:
+            node.result = None
+
+    def flattened_sub_tree_as_json(self):
+        result_list = [self.json()]
+        for child in self.children:
+            result_list += child.flattened_sub_tree_as_json()
+        return result_list
+
+    def json(self):
+        d = {}
+        d["id"] = self.id
+        d["children"] = self.children
+        d["result"] = self.result
+        d["visible_on_endpoints"] = self.visible_on_endpoints
+        d["filter_via_input_ids"] = self.filter_via_input_ids
+        return d
+
+
+HTML_ROOT = None
+
+
+def build_html_elements(specter):
+    """
+    This builds all HtmlElements that should be highlighted during a search.
+    It also encodes which functions will be used for searching.
+
+    Returns:
+        HtmlElement: This is the html_root, which has all children linked inside
+    """
+    html_root = HtmlElement(None)
+    wallets = HtmlElement(html_root, id="toggle_wallets_list")
+    devices = HtmlElement(html_root, id="toggle_devices_list")
+
+    def search_in_structure(search_term, l):
```

Comment:
Puh, this function in function-scope is imho quite horrible. Especially evil is the example below where you're not even returning a value but just changing variables which are in the scope. To me, this is a very confusing programming-model.

Shouldn't it be possible to include these methods in the HtmlElement-class?

Reply:
I agree, this needs a restructuring.